c_emulator/riscv_default_callbacks.c

```diff
@@ -0,0 +1,91 @@
+#include "riscv_config.h"
+#include <stdlib.h>
+
+void zcsr_name_map_forwards(sail_string *rop, uint64_t);
+
+static uint8_t *get_lbits_data(lbits val)
+{
+  uint8_t *data = (uint8_t *)calloc(val.len, sizeof(uint8_t));
+  mpz_export(data, NULL, -1, 1, 0, 0, val.bits);
+  return data;
+}
+
+/* Implementations of default callbacks for trace printing.
+ *
+ * The model assumes that these functions do not change the state of the model.
+ */
+int mem_write_callback(uint64_t addr, uint64_t width, lbits value)
+{
+  if (config_print_mem_access) {
+    char *lbits_data = get_lbits_data(value);
+    printf("mem[0x%.16lX] <- 0x", addr);
+    for (int i = width - 1; i >= 0; --i)
+      printf("%02hhX", lbits_data[i]);
+    printf("\n");
+    free(lbits_data);
+  }
+}
+
+int mem_read_callback(const char *type, uint64_t addr, uint64_t width,
+                      lbits value)
+{
+  if (config_print_mem_access) {
+    char *lbits_data = get_lbits_data(value);
+    printf("mem[%s,0x%.16lX] -> 0x", type, addr);
+    for (int i = width - 1; i >= 0; --i)
+      printf("%02hhX", lbits_data[i]);
+    printf("\n");
+    free(lbits_data);
+  }
+}
+
+int mem_exception_callback(uint64_t addr, uint64_t num_of_exception) { }
+
+int xreg_write_callback(unsigned reg, uint64_t value)
+{
+  if (config_print_reg)
+    printf("x%d <- 0x%.16lX\n", reg, value);
+}
+
+int freg_write_callback(unsigned reg, uint64_t value)
+{
+  /* TODO: will only print bits; should we print in floating point format? */
+  if (config_print_reg)
```

Comment:
Minor, but can you always use braces - single-statement blocks a bit of a footgun in C.

I thought we had clang-format set up to enforce that but maybe not. I'll double check that.